File: core/descriptor.loader.java/src/org/jetbrains/jet/lang/resolve/java/lazy/descriptors/LazyJavaPackageFragment.java
package org.jetbrains.jet.lang.resolve.java.lazy.descriptors;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.jet.lang.descriptors.NamespaceDescriptor;
import org.jetbrains.jet.lang.descriptors.annotations.AnnotationDescriptor;
import org.jetbrains.jet.lang.descriptors.impl.AbstractNamespaceDescriptorImpl;
import org.jetbrains.jet.lang.descriptors.impl.NamespaceDescriptorParent;
import org.jetbrains.jet.lang.resolve.name.FqName;
import org.jetbrains.jet.lang.resolve.scopes.JetScope;

import java.util.Collections;

public class LazyJavaPackageFragment extends AbstractNamespaceDescriptorImpl implements NamespaceDescriptor, LazyJavaDescriptor {
    private final FqName fqName;
    private final JetScope memberScope;

    public LazyJavaPackageFragment(
            @NotNull NamespaceDescriptorParent containingDeclaration,
            @NotNull FqName fqName,
            @NotNull JetScope memberScope
    ) {
        super(containingDeclaration, Collections.<AnnotationDescriptor>emptyList(), fqName.shortName());
        this.fqName = fqName;
        this.memberScope = memberScope;
    }

    @NotNull
    @Override
    public JetScope getMemberScope() {
        return memberScope;
    }

    @NotNull
    @Override
    public FqName getFqName() {
        return fqName;
    }
}

File: core/descriptor.loader.java/src/org/jetbrains/jet/lang/resolve/java/lazy/descriptors/LazyJavaPackageFragment.kt
package org.jetbrains.jet.lang.resolve.java.lazy.descriptors

import org.jetbrains.jet.lang.descriptors.NamespaceDescriptor
import org.jetbrains.jet.lang.descriptors.impl.AbstractNamespaceDescriptorImpl
import org.jetbrains.jet.lang.descriptors.impl.NamespaceDescriptorParent
import org.jetbrains.jet.lang.resolve.name.FqName
import org.jetbrains.jet.utils.emptyList
import org.jetbrains.jet.lang.resolve.java.JavaClassFinder
import org.jetbrains.jet.lang.resolve.lazy.storage.StorageManager

public class LazyJavaPackageFragment(
        storageManager: StorageManager,
        containingDeclaration: NamespaceDescriptorParent,
        finder: JavaClassFinder,
        private val _fqName: FqName
) : AbstractNamespaceDescriptorImpl(
        containingDeclaration,
        emptyList(),
        _fqName.shortName()
    ), NamespaceDescriptor, LazyJavaDescriptor {

    private val _memberScope = LazyJavaPackageFragmentScope(this, storageManager, finder)

    override fun getMemberScope() = _memberScope
    override fun getFqName() = _fqName
}
